### PR TITLE
`AccountType`, add `EthImplicitAccount`

### PR DESCRIPTION
**Pull request by staffik**
Part of https://github.com/near/nearcore/issues/10018.

This PR introduces some changes from https://github.com/near/nearcore/pull/9969 laying groundwork for real protocol changes to be done in a separate PR.

Summary:
- Add `AccountType` enum: `NamedAccount`, `NearImplicitAccount`, or `EthImplicitAccount`.
- Parse 40 characters long hexadecimal addresses prefixed with `'0x'` as `EthImplicitAccount`.
- For now, `AccountType::is_implicit()` returns false for `EthImplicitAccount`.

### CODE REVIEW

**Comment by wacban** [comment]
I didn't even know about this repo. Where is it used? Do we need to keep it synchronized with nearcore?

**Review by wacban — DISMISSED** [review]
LGTM

**Review by birchmd — DISMISSED** [review]
I also do not know what this repo is used for, but it seems good to keep it up to date with nearcore.

**Comment by staffik** [comment]
@frol Can you look at this PR? How to make a release so that it is accessible from nearcore?

**Comment by frol** [comment]
> I didn't even know about this repo. Where is it used? Do we need to keep it synchronized with nearcore?

It has to be used everywhere in Rust code in NEAR ecosystem:

* https://github.com/near/nearcore/pull/10019
* https://github.com/near/near-sdk-rs/issues/1105

near-account-id has been part of nearcore, but since we wanted to stabilize it and have 1.0 release, we extracted it into a standalone repo.

It is all part of a major stabilization effort:

* [borsh-rs 1.0 is released](https://github.com/near/borsh-rs/releases/tag/borsh-v1.0.0)
* [near-token-rs](https://crates.io/crates/near-token) is nice and small, and will probably be promoted to 1.0 soon
* [near-gas-rs](https://crates.io/crates/near-gas) is nice and small, and will probably be promoted to 1.0 soon
* [near-sdk-rs 5.0 release is coming](https://github.com/near/near-sdk-rs/issues/946)
* [near-workspaces-rs also targeting 1.0](https://github.com/near/near-workspaces-rs/issues/113)

In order to release 1.0, we need to ensure that the underlying dependencies are stable, so near-account-id, was extracted, and near-primitives-core might also be extracted at a certain point (though, I won't even try to convince nearcore to use the primitives, so probably that will be a standalone effort)

**Comment by wacban** [comment]
@frol Thanks for the explanation. As nearcore developers we need a way to iterate quickly on the repo. I don't know about 1.0 but need to get that PR in, released and synced to nearcore quickly. Can you confirm that we are free to work on this repo as we would on nearcore (e.i. the same review process), please? Can you share instructions on how to release and sync to nearcore please? Ideally in form of some documentation but a comment here would be a good start too.

File: src/account_id_ref.rs [hunk 1]
```diff
@@ -29,6 +29,28 @@ use crate::{AccountId, ParseAccountError};
...
+impl AccountType {
+    pub fn is_implicit(&self) -> bool {
+        match &self {
+            Self::NearImplicitAccount => true,
+            // TODO(eth-implicit) change to true later, see https://github.com/near/nearcore/issues/10018
+            Self::EthImplicitAccount => false,
+            Self::NamedAccount => false,
+        }
+    }
+}
```
Comment (frol) [inline src/account_id_ref.rs:52]: What is the motivation for this extra enum? I feel that `is_eth_implicit` and `is_near_implicit` + `is_implicit` would be enough.

File: src/account_id_ref.rs [hunk 2]
```diff
@@ -150,21 +195,31 @@ impl AccountIdRef {
...
+    /// assert!(!alice.is_near_implicit());
     ///
     /// let rando = "98793cd91a3f870fb126f66285808c7e094afcfc4eda8a970f6648cdf0dbd6de"
     ///     .parse::<AccountId>()
     ///     .unwrap();
-    /// assert!(rando.is_implicit());
+    /// assert!(rando.is_near_implicit());
     /// ```
-    pub fn is_implicit(&self) -> bool {
+    pub fn is_near_implicit(&self) -> bool {
```
Comment (frol) [inline src/account_id_ref.rs:205]: I would consider keeping `is_implicit` (= `is_near_implicit || is_eth_implicit`) to denote that account does not need to be created explicitly with `CreateAccount` action. What do you think?

**Review by frol — CHANGES_REQUESTED** [review]

**Comment by frol** [comment]
> As nearcore developers we need a way to iterate quickly on the repo. I don't know about 1.0 but need to get that PR in, released and synced to nearcore quickly.

Well, if you break the ecosystem fast it does not really help, so I encourage nearcore team to take it into consideration as well. near-account-id is the primitive that needs to be the same across the board.

> Can you confirm that we are free to work on this repo as we would on nearcore (e.i. the same review process), please?

It is slightly different here due to the focus on delivering a reusable crate (well, it should have been treated this way in nearcore, but it was not). The difference is that we need to make a conscious decision about breaking changes (in nearcore only protocol changes were considered breaking).

> Can you share instructions on how to release and sync to nearcore please?

It is easy - after this PR is merged to main branch, automation will kick in, and create a release PR, once that PR is reviewed and merged in main branch, automation will take care of publishing the release.

File: src/account_id_ref.rs (hunk 1 above)
Comment (wacban) [reply src/account_id_ref.rs:52]: @frol please see the links in the pr description for more details about this project

The reason for using an enum here is good software development practice. We want to be explicit about the different types of accounts and force ourselves and future developers to be aware and correctly handle the different types of account. Additionally this is more future proof if we ever need to add more account types. This was also discussed on the original PR in nearcore.

File: src/account_id_ref.rs (hunk 2 above)
Comment (wacban) [reply src/account_id_ref.rs:205]: Please see the original pr for more details. 

TLDR - we want to be explicit about type different types of implicit account. In order to not impede developer productivity we are adding the is_implict method to the account type enum so it is still reasonably convenient to use while making developers aware of the different account types.

File: src/account_id_ref.rs (hunk 1 above)
Comment (frol) [reply src/account_id_ref.rs:52]: > We want to be explicit about the different types of accounts and force ourselves and future developers to be aware and correctly handle the different types of account.

If that is the intent, we should remove the `is_near_implicit` and `is_eth_implicit` methods from the AccountIdRef type as having two different ways to do the same thing is not helping here and does not really force users to handle all the cases.

Just for the future reference, I will link the prior discussion: https://github.com/near/nearcore/pull/9969#discussion_r1367162654

**Comment by wacban** [comment]
@frol Thanks! 

> Well, if you break the ecosystem fast it does not really help, so I encourage nearcore team to take it into consideration as well.  

I don't think breaking the ecosystem at any speed is helpful but I would argue that breaking it fast is better than breaking is slow. Then at least we can recover equally fast. 

> near-account-id is the primitive that needs to be the same across the board. 

That makes sense but it's still not clear to me what is the process for making changes to this new repo. In particular this new feature will be released with a protocol version upgrade while this repo doesn't seem to be aware of it. 

JFYI this PR should be mostly a no-op, preparatory work. In a latter PR @staffik will do the work described in TODOs (make is_implicit return true for the ethereum account types). 

> The difference is that we need to make a conscious decision about breaking changes (in nearcore only protocol changes were considered breaking). 

Can you describe this process please? This project will be a breaking change so we need to know how to handle it. 

>  automation will take care of publishing the release. 

Brilliant, thanks!

File: src/account_id_ref.rs (hunk 1 above)
Comment (frol) [reply src/account_id_ref.rs:52]: P.S. Given that "renaming" `is_implicit` to `is_near_implicit` is already a breaking change, removing all those `is*implicit` methods might be just fine.

Just to model the usage here:

BEFORE

```rust
fn send_near(receiver_id: AccountIdRef) {
    if !receiver_id.is_implicit() {
        assert!(is_account_recorded_on_chain(receiver_id), "you cannot send tokens to non-existing non-implicit accounts");
    }
    ...
}
```

AFTER renaming to `is_near_implicit` and naive changes:

```rust
fn send_near(receiver_id: AccountIdRef) {
    if !receiver_id.is_near_implicit() {
        assert!(is_account_recorded_on_chain(receiver_id), "you cannot send tokens to non-existing non-implicit accounts");
    }
    ...
}
```

AFTER if no `is_*implicit` methods:

```rust
fn send_near(receiver_id: AccountIdRef) {
    if let AccountType::NamedAccount = receiver_id.get_account_type() {
        assert!(is_account_recorded_on_chain(receiver_id), "you cannot send tokens to non-existing non-implicit accounts");
    }
    ...
}
```

OR

```rust
fn send_near(receiver_id: AccountIdRef) {
    match receiver_id.get_account_type() {
        AccountType::NamedAccount => {
            assert!(is_account_recorded_on_chain(receiver_id), "you cannot send tokens to non-existing non-implicit accounts");
        }
        AccountType::NearImplicitAccount => {},
        AccountType::EthImplicitAccount => {},
    }
    ...
}
```

Comment (staffik) [reply src/account_id_ref.rs:52]: @frol Would it be ok if I just rename `pub fn is_eth_implicit` to `fn is_eth_implicit` and `pub fn is_near_implicit` to `fn is_near_implicit`? I would remove docstrings with tests from `is_eth_implicit` and `is_near_implicit`, and add corresponding docstring with tests for `get_account_type`.

Comment (frol) [reply src/account_id_ref.rs:52]: @staffik Do you mean removing `pub` from `is_eth_implicit` and `is_near_implicit`? I would consider moving them to `validation.rs` file

Comment (staffik) [reply src/account_id_ref.rs:52]: @frol Yes. Sure, I will try to move them to `validation.rs` file.

File: src/account_id_ref.rs [hunk 3]
```diff
@@ -29,6 +29,28 @@ use crate::{AccountId, ParseAccountError};
 #[cfg_attr(feature = "abi", derive(schemars::JsonSchema, BorshSchema))]
 pub struct AccountIdRef(pub(crate) str);
 
+/// Enum representing possible types of accounts.
```
[the remark this answers is not in the document]
Comment (staffik) [reply src/account_id_ref.rs:32]: @frol Done

File: src/account_id_ref.rs [hunk 4]
```diff
@@ -29,6 +29,33 @@ use crate::{AccountId, ParseAccountError};
...
+    /// An account which address starts with '0x', followed by 40 hex characters.
+    EthImplicitAccount,
+}
+
+impl AccountType {
+    pub fn is_implicit(&self) -> bool {
+        match &self {
+            Self::NearImplicitAccount => true,
+            // TODO(eth-implicit) change to true later, see https://github.com/near/nearcore/issues/10018
+            Self::EthImplicitAccount => false,
```
Comment (frol) [inline src/account_id_ref.rs:53]: This would look like a breaking change for near-account-id crate, so I would just return `true` here from the get-go. Any strong objections? It would be a separate challenge for the future if NEAR Protocol changes account-id rules drastically.

**Review by frol — APPROVED** [review]
Overall, looks good to me. Just one discussion point regarding `is_implicit`, see below.

File: src/account_id_ref.rs (hunk 4 above)
Comment (staffik) [reply src/account_id_ref.rs:53]: Ok, I can get around that within the nearcore repo, to keep this repo stable.

Comment (wacban) [reply src/account_id_ref.rs:53]: SGTM

Comment (wacban) [reply src/account_id_ref.rs:53]: We still need to work out what to do about the existing named accounts that have the eth account format. I would still go ahead with this PR as agreed because it's blocking everything elsewhere.

**Comment by staffik** [comment]
@frol or @wacban, I need approve again because I pushed a commit now.

**Review by wacban — APPROVED** [review]

**Comment by frol** [comment]
@staffik @wacban 1.0.0-alpha.2 is [published](https://crates.io/crates/near-account-id/1.0.0-alpha.2)!
